[SelectionDAG] Deal with POISON for INSERT_VECTOR_ELT/INSERT_SUBVECTOR

bjope opened the pull request:
As reported in https://github.com/llvm/llvm-project/issues/141034 SelectionDAG::getNode had some unexpected
behaviors when trying to create vectors with UNDEF elements. Since
we treat both UNDEF and POISON as undefined (when using isUndef())
we can't just fold away INSERT_VECTOR_ELT/INSERT_SUBVECTOR based on
isUndef(), as that could make the resulting vector more poisonous.

Same kind of bug existed in DAGCombiner::visitINSERT_SUBVECTOR.

Here are some examples:

This fold was done even if vec[idx] was POISON:
  INSERT_VECTOR_ELT vec, UNDEF, idx -> vec

This fold was done even if any of vec[idx..idx+size] was POISON:
  INSERT_SUBVECTOR vec, UNDEF, idx -> vec

This fold was done even if the elements not extracted from vec could
be POISON:
  sub = EXTRACT_SUBVECTOR vec, idx
  INSERT_SUBVECTOR UNDEF, sub, idx -> vec

With this patch we avoid such folds unless we can prove that the
result isn't more poisonous when eliminating the insert.

Fixes https://github.com/llvm/llvm-project/issues/141034

RKSimon commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -22915,14 +22916,29 @@ SDValue DAGCombiner::visitINSERT_VECTOR_ELT(SDNode *N) {
+    } else if (DAG.isGuaranteedNotToBePoison(InVec)) {
+      return InVec;
+    }
   }
```
Still not sure why we can't do this:
```cpp
  if (InVal.isUndef()) {
    if (IndexC && VT.isFixedLengthVector()) {
      APInt EltMask = APInt::getOneBitSet(VT.getVectorNumElements(),
                                          IndexC->getZExtValue());
      if (DAG.isGuaranteedNotToBePoison(InVec, EltMask))
        return InVec;
    }
    return DAG.getFreeze(InVec);
  }
```

arsenm replied:
> (Juggling with stacked pull requests in github is still something that I'm still not very confortable with. It really messes up my local workflow as I can't just "git rebase -i" to move between the patch stack.)

You need to use one of the tools (graphite or spr), these manually stacked with pre-commits are as much of a pain to review as to submit

RKSimon commented on llvm/test/CodeGen/X86/any_extend_vector_inreg_of_broadcast.ll:
```diff
@@ -4716,6 +4717,7 @@ define void @vec384_i64_widen_to_i128_factor2_broadcast_to_v3i128_factor3(ptr %i
 ; AVX-NEXT:    vmovdqa %xmm3, 32(%rcx)
 ; AVX-NEXT:    vmovdqa %xmm1, 48(%rcx)
 ; AVX-NEXT:    vmovdqa %xmm2, (%rcx)
+; AVX-NEXT:    vzeroupper
```
These look like merge errors as I've recently updated x86 shuffles to stop emitting so many vzeroupper ops

bjope commented:
I'm still looking for advice on how to move forward with this.

- Add more reviewers?
- Re-order patches to avoid temporary test regressions (i.e. do fixups first)?  (problem here is I would need to find motivating test cases)?
- Squash everything into a mega patch?
- Split up in even smaller patches?
- Do I need to get all patches (including #143103 and #143105) approved and landed simultaneously?


As I see it fixing #141034 is needed to move forward with the fixes for #138513 (as they rely on that insert of UNDEF isn't surprisingly turned into inserting POISON). I've been stalling on updating the patches for #138513 while waiting for this patch to land.

RKSimon commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -27560,18 +27575,42 @@ SDValue DAGCombiner::visitINSERT_SUBVECTOR(SDNode *N) {
+  // result more poisonous).
+  if (N1.isUndef()) {
+    if (N1.getOpcode() == ISD::POISON)
+      return N0;
```
why don't we do this for the visitINSERT_VECTOR_ELT case?

RKSimon commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -27560,18 +27575,42 @@ SDValue DAGCombiner::visitINSERT_SUBVECTOR(SDNode *N) {
+        if (DAG.isGuaranteedNotToBePoison(N1.getOperand(0), ~EltMask))
+          return N1.getOperand(0);
+      } else if (DAG.isGuaranteedNotToBePoison(N1.getOperand(0)))
+        return N1.getOperand(0);
```
`return DAG.getFreeze(N1.getOperand(0))`?

bjope replied:
I know that I tried that, but that resulted in regressions that I couldn't solve easily. Right now I don't remember the exact scenario. Probably something with the scalable vectors, and a FREEZE that couldn't be eliminated and thereby blocking some combine.

RKSimon commented on llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:
```diff
@@ -7923,23 +7923,42 @@ SDValue SelectionDAG::getNode(unsigned Opcode, const SDLoc &DL, EVT VT,
+          return N1;
+      } else if (isGuaranteedNotToBePoison(N1))
+        return N1;
+    }
```
How many regressions do we get it we drop all this from getNode() and just leave the combine to do it?

We've often ended up doing more then basic canonicalisation in getNode() just to avoid one specific test regression.....

RKSimon commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -27560,18 +27575,42 @@ SDValue DAGCombiner::visitINSERT_SUBVECTOR(SDNode *N) {
-    return N0;
+  // If inserting an UNDEF, just return the original vector (unless it makes the
+  // result more poisonous).
+  if (N1.isUndef()) {
```
I keep forgetting than SDValue::isUndef() is really SDValue::isUndefOrPoison() :/

RKSimon commented on llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:
```diff
@@ -7968,11 +7987,37 @@ SDValue SelectionDAG::getNode(unsigned Opcode, const SDLoc &DL, EVT VT,
+          return N1;
+      } else if (isGuaranteedNotToBePoison(N1))
+        return N1;
+    }
```
again - are we doing too much in getNode at this stage?

RKSimon reviewed (commented):
I've taken a look - stretching everything across patches does it make it trickier, but stacked PRs aren't any better

RKSimon commented:
After #145939 I tried to create a first step to help with this on #146332 but to address regressions its starting to look at lot like this patch tbh - please can you rebase and help identify the missing opcodes we need to add to the canCreateUndefOrPoison callbacks?

bjope commented:
So right now, if checking out https://github.com/llvm/llvm-project/pull/143105 (the branch users/bjope/insertundef_4), the only test cases that has a diff compared to the base commit from main are these five:
```
llvm/test/CodeGen/AArch64/arm64-build-vector.ll
llvm/test/CodeGen/AArch64/concat-vector-add-combine.ll
llvm/test/CodeGen/AArch64/sve-fixed-length-frame-offests.ll
llvm/test/CodeGen/X86/pr62286.ll
llvm/test/CodeGen/X86/vector-shuffle-combining.ll
```

At least a couple of those seem trivial. Maybe the last one would be the main concern if any?

RKSimon commented:
One thing that is hurting us is the freeze getting in between a load and ext prevent a extload forming:
```
t13: i8,ch = load<(load (s8) from %ir.p2)> t0, t6, undef:i64
t47: i8 = freeze t13
t14: i16 = sign_extend t47
```
I'm not certain but it'd be great if we were allowed to fold these to:
```
t13: i16,ch = load<(load (s8) from %ir.p2), sext from i8> t0, t6, undef:i64
t47: i16 = freeze t13
t14: i16 = assertsext t47, MVT::i8
```
WDYT?

bjope commented:
> Similarly - should we be allowed to remove freezes from constant pool loads entirely (maybe with some handling for undef elts in the Constant)?

I think so, but not quite sure how to best detect that in a good way.

One idea I had was to check for `TLI->getTargetConstantFromLoad(cast<LoadSDNode>(Op))` in `isGuaranteedNotToBeUndefOrPoison` (or `canCreateUndefOrPoison`) to identify LOAD from constant pool. But I haven't seen any benefit of that (except maybe in one lit test). That target hook is also only implemented for two targets.

Another thinking is that other constant loads also could be seen as frozen (e.g. loading from fixed stack). If so, then something like this in `isGuaranteedNotToBeUndefOrPoison` might be used (if it can be guaranteed that the LOAD won't result in poison due to index oob or something):
```
  case ISD::LOAD: {
    if (const PseudoSourceValue *PSV = cast<LoadSDNode>(Op)->getMemOperand()->getPseudoValue())
      if (PSV->isConstant(&MF->getFrameInfo()))
        return true;
    break;
  }
```

That latter would depend on that the MachineMemOperands are setup to identify that we load from constant pool etc.

bjope committed:
[SelectionDAG] Deal with POISON for INSERT_VECTOR_ELT/INSERT_SUBVECTOR

As reported in #141034 SelectionDAG::getNode had some unexpected
behaviors when trying to create vectors with UNDEF elements. Since
we treat both UNDEF and POISON as undefined (when using isUndef())
we can't just fold away INSERT_VECTOR_ELT/INSERT_SUBVECTOR based on
isUndef(), as that could make the resulting vector more poisonous.

Same kind of bug existed in DAGCombiner::visitINSERT_SUBVECTOR.

Here are some examples:

This fold was done even if vec[idx] was POISON:
  INSERT_VECTOR_ELT vec, UNDEF, idx -> vec

This fold was done even if any of vec[idx..idx+size] was POISON:
  INSERT_SUBVECTOR vec, UNDEF, idx -> vec

This fold was done even if the elements not extracted from vec could
be POISON:
  sub = EXTRACT_SUBVECTOR vec, idx
  INSERT_SUBVECTOR UNDEF, sub, idx -> vec

With this patch we avoid such folds unless we can prove that the
result isn't more poisonous when eliminating the insert.

Fixes #141034

bjope commented:
Ping!

Would be really nice to move forward with this patch somehow, to finally get back to fixing the miscompiles we saw in https://github.com/llvm/llvm-project/issues/138513 (which is where I started off 3-4 months ago).

I don't know if the remaining diffs are acceptable now?

The aarch/sve and riscv/rvv diffs can be resolved by using ISD::POISON instead of ISD::UNDEF in convertToScalableVector. We could perhaps do that as a follow-up (to get a motivating test case...).

The diff here https://github.com/llvm/llvm-project/pull/143102/files#diff-3f4fe0b0594906b6df9b5eef955cf7fbc242076e17787dbc845f2adc40ee8d8f is perhaps a bit unfortunate. But I don't have a good solution. However, the IR is not expected to look like that when entering llc, since instcombine would get rid of the insertelement instructions that insert undef. I think the situation was similar for the X86 (SpinningCube) test case that show a diff.

RKSimon commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -23308,14 +23309,28 @@ SDValue DAGCombiner::visitINSERT_VECTOR_ELT(SDNode *N) {
+      if (DAG.isGuaranteedNotToBePoison(InVec, EltMask))
+        return InVec;
+    }
+    return DAG.getFreeze(InVec);
```
How much of this do you think is still necessary if we can get #152107 done? Would that allow us to just freeze more aggressively, and not try to handle so many isGuaranteedNotToBePoison cases locally and leave visitFREEZE to handle it?

bjope replied:
Right. We can probably be more aggressive and always freeze the input vector when inserting ISD::UNDEF.

Seems like it is enough to add ISD::INSERT_SUBVECTOR to AllowMultipleMaybePoisonOperands in visitFREEZE to, and then we can just freeze the input vector instead of inserting UNDEF (both to INSERT_VECTOR_ELT and INSERT_SUBVECTOR). That is when considering the existing lit tests, as I guess there might be other scenarios not covered by lit tests.

RKSimon reviewed (commented):
What happened to the plan to to add ISD::INSERT_SUBVECTOR to the multiple operand list in visitFREEZE? With that in place much of this logic can be simplified and we just FREEZE instead of calling isGuaranteedNotToBePoison explicitly so much

RKSimon commented on llvm/test/CodeGen/X86/vector-shuffle-combining.ll:
```diff
@@ -3572,45 +3572,53 @@ define void @SpinningCube() {
+; SSE2-NEXT:    shufps {{.*#+}} xmm0 = xmm0[0,0,0,0]
+; SSE2-NEXT:    mulps {{\.?LCPI[0-9]+_[0-9]+}}(%rip), %xmm0
+; SSE2-NEXT:    addps %xmm2, %xmm0
+; SSE2-NEXT:    movaps %xmm0, (%rax)
```
any luck finding the cause of this regression? are we stuck trying to freeze a target shuffle opcode?

bjope replied:
To clarify a bit. InstCombine is reducing this to:
```
define void @SpinningCube() {
entry:
  store float 1.000000e+00, ptr undef, align 4
  ret void
}
```
Kind of indicating that the backend is pretty bad at simplifying this already before this patch.

Main diff before legalization is that we get
```
t55: f32 = freeze undef:f32
  t56: v4f32 = BUILD_VECTOR t55, t55, t55, ConstantFP:f32<1.000000e+00>
```
instead of
```
t49: v4f32 = BUILD_VECTOR undef:f32, undef:f32, undef:f32, ConstantFP:f32<1.000000e+00>
```
The latter would be legalized as a load from constant pool. But I don't think that happens when we have the BUILD_VECTOR with frozen undef. Instead the legalizer introduce another vector_shuffle(?).
So we get
```
Legalizing: t56: v4f32 = BUILD_VECTOR t55, t55, t55, ConstantFP:f32<1.000000e+00>
Trying custom legalization
Successfully custom legalized node
 ... replacing: t56: v4f32 = BUILD_VECTOR t55, t55, t55, ConstantFP:f32<1.000000e+00>
     with:      t74: v4f32 = vector_shuffle<4,5,6,3> t72, t73
```
instead of
```
Legalizing: t49: v4f32 = BUILD_VECTOR undef:f32, undef:f32, undef:f32, ConstantFP:f32<1.000000e+00>
Trying custom legalization
Could not custom legalize node
Trying to expand node
Successfully expanded node
 ... replacing: t49: v4f32 = BUILD_VECTOR undef:f32, undef:f32, undef:f32, ConstantFP:f32<1.000000e+00>
     with:      t67: v4f32,ch = load<(load (s128) from constant-pool)> t0, ConstantPool:i64<<4 x float> <float undef, float undef, float undef, float 1.000000e+00>> 0, undef:i64
```

RKSimon reviewed (approved):
LGTM - you're starting to hit the same problems as I am with #152107 - in particular when we can remove freezes from constant loads (or freezing large loads that need to be split), but this patch isn't making this any tougher to handle in the future

bjope committed:
Merge branch 'main' into users/bjope/insertundef_2
